tools/serverpod_cli/lib/src/serverpod_packages_version_check/serverpod_packages_version_check.dart

```
@@ -18,102 +19,115 @@ class ServerpodPackagesVersionCheckWarnings {
 }
 
-List<SourceSpanSeverityException> performServerpodPackagesAndCliVersionCheck(
-  Version cliVersion,
-  GeneratorConfig config,
-) {
-  List<SourceSpanSeverityException> accumulatedWarnings = [];
-
-  var pubspecFiles = <File>[
-    File('pubspec.yaml'), // Server pubspec file
-    File(path.joinAll(config.clientPackagePathParts)), // Client pubspec file
-  ];
-  for (var pubspecFile in pubspecFiles) {
-    var file = pubspecFile.readAsStringSync();
-    var pubspec = Pubspec.parse(file);
-    var yaml = loadYaml(file, sourceUrl: pubspecFile.uri) as YamlMap;
-    var warnings = _validateServerpodPackagesVersion(cliVersion, pubspec, yaml);
-
-    accumulatedWarnings.addAll(warnings);
+/// A pubspec plus the yaml map it was read from, and a list of all augmented
+/// dependencies regardless of kind.
+class PubspecPlus {
+  final Pubspec pubspec;
+  final YamlMap yaml;
+
+  PubspecPlus.fromFile(File file) : this.parse(file.readAsStringSync());
+
+  PubspecPlus.parse(String yamlString)
+      : pubspec = Pubspec.parse(yamlString),
+        yaml = loadYaml(yamlString) as YamlMap; // safe as it is already parsed
+
+  /// All dependencies in the pubspec, regardless of type. Augmented with name,
+  /// kind, and span.
+  // Only calculate once, but no need to calculate unless needed
+  late List<Dep> deps = _iteratedDeps().toList();
+
+  Iterable<Dep> _iteratedDeps() sync* {
+    for (var x in [
+      (
+        deps: pubspec.dependencies,
+        type: DepKind.normal,
+        yaml: yaml['dependencies'] as YamlMap?
+      ),
+      (
+        deps: pubspec.devDependencies,
+        type: DepKind.dev,
+        yaml: yaml['dev_dependencies'] as YamlMap?
+      ),
+      (
+        deps: pubspec.dependencyOverrides,
+        type: DepKind.override,
+        yaml: yaml['dependency_overrides'] as YamlMap?
+      ),
+    ]) {
+      for (var e in x.deps.entries) {
+        var span = x.yaml!.nodes[e.key]!.span; // ! safe as we loop over entries
```

Suggestion: We should always avoid `!` even if we know it's safe.

Please fetch the value using `?` and throw an explicit exception if the value is null.

---

What exception do we throw, when it would be an internal bug if a situation ever happens?

---

Sorry, it's not an exception; we should throw an error since it is not actionable from the user. Perhaps a `StateError` with a message explaining the unexpected condition we are in?